plugin-server/src/cdp/cdp-processed-events-consumer.ts

```diff
@@ -0,0 +1,258 @@
+
+                                    if (
+                                        await this.organizationManager.hasAvailableFeature(
+                                            clickHouseEvent.team_id,
+                                            'group_analytics'
+                                        )
+                                    ) {
+                                        // If the organization has group analytics enabled then we enrich the event with group data
+                                        groupTypes = await this.groupTypeManager.fetchGroupTypes(
+                                            clickHouseEvent.team_id
+                                        )
+                                    }
+
+                                    // TODO: Clean up all of this and parallelise
+                                    // TODO: We can fetch alot of teams and things in parallel
+
+                                    const team = await this.teamManager.fetchTeam(clickHouseEvent.team_id)
+                                    if (!team) {
+                                        return
+                                    }
+                                    const globals = convertToHogFunctionInvocationContext(
+                                        clickHouseEvent,
+                                        team,
+                                        this.config.SITE_URL ?? 'http://localhost:8000',
+                                        groupTypes
+                                    )
+
+                                    invocations.push({
+                                        globals,
+                                    })
+                                } catch (e) {
+                                    status.error('Error parsing message', e)
+                                }
+                            })
+                        )
+                    },
+                })
+                heartbeat()
+
+                const invocationResults: HogFunctionInvocationResult[] = []
+
+                await runInstrumentedFunction({
+                    statsKey: `cdpFunctionExecutor.handleEachBatch.consumeBatch`,
+                    func: async () => {
+                        // TODO: Parallelise this
+                        for (const message of invocations) {
+                            const results = await this.consume(message)
+                            invocationResults.push(...results)
+                            heartbeat()
+                        }
+                    },
+                })
+
+                // TODO: Follow up - process metrics from the invocationResults
+                // await runInstrumentedFunction({
+                //     statsKey: `cdpFunctionExecutor.handleEachBatch.queueMetrics`,
+                //     func: async () => {
+                //         // TODO:
+                //     },
+                // })
```

Comment:
Thinking through the `TODO, parallelise this`, we want to get to a point where this entire loop is `sync`, and returns a result with just a few arrays like `{ eventsToClickhouse: [], invocationsToWebhooks: [], invocationsToSleepyHog: [] }`, and then runs them in batches.

Reply:
Yeah it will depend a bit on a few factors but its not a biggy